fix concurrent synchorinzation problem in NodeChannel::try_send_batch

dataroaring opened the pull request:
The patch fixes two problems.
1. Memory order problem accessing _last_patch_processed_finished and in_flight, actually _last_patch_processed_finished is redudant, so the patch removes it.
2. synchorization in join on cid.

Fix for https://github.com/apache/incubator-doris/issues/8725.

# Proposed changes

Issue Number: close #8725 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
4. Has unit tests been added: (Yes/No/No Need)
5. Has document been added or modified: (Yes/No/No Need)
6. Does it need to update dependencies: (Yes/No)
7. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

dataroaring committed:
fix concurrent synchorinzation problem in NodeChannel::try_send_batch

The detailed message is put in tablet_sink.cpp as comment.

Fix for https://github.com/apache/incubator-doris/issues/8725.

morningman commented on be/src/exec/tablet_sink.cpp:
```diff
@@ -458,8 +458,37 @@ int NodeChannel::try_send_and_fetch_status(RuntimeState* state,
+    // |----------------------------------------------------------------------------
+    // | Thread try_send_and_fetch_status  | Thread try_send_batch                  |
+    // |-----------------------------------------------------------------------------
+    // | read packet_in_flight false       | before seting packet_in_flight         |
```
```suggestion
    // | read packet_in_flight false       | before setting packet_in_flight         |
```

morningman replied:
So this may trigger `DCHECK(!_pending_batches.empty());` in `NodeChannel::try_send_batch`?

dataroaring replied:
yep

morningman reviewed (dismissed):
LGTM

dataroaring commented:
please do not merge this patch, the patch brings other problems.

dataroaring committed:
Recover last_batch_process_finished after failing to submit task

dataroaring committed:
fix typo

dataroaring committed:
simplify logic assuring one rpc is in flight in OlapTableSink

We should just remove _last_patch_processed_finished, which
complicates the logic. packet_in_flight is enough to do so.

dataroaring committed:
assure join not return before rpc ends

cid is not synchronized, so it is not safe to use it for joining.

We should use in_flight as the only one flag indicating a rpc is
running.

dataroaring commented:
Please do not merge this patch, I have not tested it thoroughly.

morningman reviewed (approved):
LGTM
